rclcpp/include/rclcpp/any_subscription_callback.hpp

```diff
@@ -15,256 +15,338 @@
 #ifndef RCLCPP__ANY_SUBSCRIPTION_CALLBACK_HPP_
 #define RCLCPP__ANY_SUBSCRIPTION_CALLBACK_HPP_
 
-#include <rmw/types.h>
-
 #include <functional>
 #include <memory>
 #include <stdexcept>
 #include <type_traits>
 #include <utility>
+#include <variant>
+
+#include "tracetools/tracetools.h"
+#include "tracetools/utils.hpp"
 
 #include "rclcpp/allocator/allocator_common.hpp"
+#include "rclcpp/detail/subscription_callback_type_helper.hpp"
 #include "rclcpp/function_traits.hpp"
 #include "rclcpp/message_info.hpp"
-#include "rclcpp/visibility_control.hpp"
-#include "tracetools/tracetools.h"
-#include "tracetools/utils.hpp"
+
+template<class> inline constexpr bool always_false_v = false;
 
 namespace rclcpp
 {
 
-template<typename MessageT, typename Alloc>
-class AnySubscriptionCallback
+namespace detail
+{
+
+template<typename MessageT, typename AllocatorT>
+struct MessageDeleterHelper
 {
-  using MessageAllocTraits = allocator::AllocRebind<MessageT, Alloc>;
+  using MessageAllocTraits = allocator::AllocRebind<MessageT, AllocatorT>;
   using MessageAlloc = typename MessageAllocTraits::allocator_type;
   using MessageDeleter = allocator::Deleter<MessageAlloc, MessageT>;
-  using ConstMessageSharedPtr = std::shared_ptr<const MessageT>;
-  using MessageUniquePtr = std::unique_ptr<MessageT, MessageDeleter>;
+};
 
-  using SharedPtrCallback = std::function<void (const std::shared_ptr<MessageT>)>;
-  using SharedPtrWithInfoCallback =
-    std::function<void (const std::shared_ptr<MessageT>, const rclcpp::MessageInfo &)>;
-  using ConstSharedPtrCallback = std::function<void (const std::shared_ptr<const MessageT>)>;
-  using ConstSharedPtrWithInfoCallback =
-    std::function<void (const std::shared_ptr<const MessageT>, const rclcpp::MessageInfo &)>;
-  using UniquePtrCallback = std::function<void (MessageUniquePtr)>;
+template<typename MessageT, typename AllocatorT>
+struct AnySubscriptionCallbackHelper
+{
+  using MessageDeleter = typename MessageDeleterHelper<MessageT, AllocatorT>::MessageDeleter;
+
+  using ConstRefCallback =
+    std::function<void (const MessageT &)>;
+  using ConstRefWithInfoCallback =
+    std::function<void (const MessageT &, const rclcpp::MessageInfo &)>;
+
+  using UniquePtrCallback =
+    std::function<void (std::unique_ptr<MessageT, MessageDeleter>)>;
   using UniquePtrWithInfoCallback =
-    std::function<void (MessageUniquePtr, const rclcpp::MessageInfo &)>;
+    std::function<void (std::unique_ptr<MessageT, MessageDeleter>, const rclcpp::MessageInfo &)>;
+
+  // TODO(wjwwood): should we drop this one, because you can achieve it easily
+  //   with the const shared_ptr & version?
+  using SharedConstPtrCallback =
+    std::function<void (std::shared_ptr<const MessageT>)>;
+  using SharedConstPtrWithInfoCallback =
+    std::function<void (std::shared_ptr<const MessageT>, const rclcpp::MessageInfo &)>;
```

Comment:
RFC

Reply:
If we are going to keep `const shared_ptr<const <MessageT> &`, then I think we should definitely keep this one too.
This signature makes the intent that a refcount "will" be kept versus "might" be kept.

Reply:
Maybe, but on the other hand, it doesn't matter to the middleware which is happening, it would boil down to something like:

```c++
shared_ptr<const MessageT> message_to_be_delivered;
if (shared_ptr<const MessageT> || const shared_ptr<const Message T> &) {
  user_callback(message_to_be_delivered);
}
```

Also, just because the user has a callback is using `shared_ptr<const MessageT>` doesn't mean they'll keep it beyond the scope of the function, and if they're not necessarily going to do that, `const shared_ptr<const MessageT> &` is slight more efficient (in theory).

But thanks for giving you preference, I'll take into consideration.

Reply:
I figured we were introducing the `const shared_ptr &` signature to better express intent of the user. If I see a user callback accepting a shared_ptr then my first thought would be that they are going to keep a reference. Otherwise, they should use a different signature.

I don't feel too strongly about if we keep this one or not, but it would be nice to document why someone should prefer one signature over another.

Reply:
We decided to keep both the `shared_ptr<const MessageT>` and `const shared_ptr<const MessageT> &` signatures.